pom.xml

```diff
@@ -49,10 +49,12 @@
         <commons-httpclient.version>3.0.1</commons-httpclient.version>
         <dep.hadoop1.version>1.2.0</dep.hadoop1.version>
         <dep.hadoop2.version>2.6.0</dep.hadoop2.version>
+        <dep.dropwizard.version>3.1.0</dep.dropwizard.version>
     </properties>
 
     <modules>
         <module>rubix-spi</module>
+        <module>rubix-common</module>
```

Comment:
Can you point to resources that you used to learn about codahale and come up with this way of adding dropwizard metrics ? Is there a doc in dropwizards-metrics project or did you copy the style from another project ? HS2 probably ?

Reply:
I am used to the pattern suggested http://metrics.dropwizard.io/3.1.0/getting-started/ and in http://metrics.dropwizard.io/3.1.0/getting-started/#reporting-via-jmx

Reply:
I am confused by the pattern and the need for `Metrics`, `CodahaleMetrics` and `MetricsFactory` class. This is the pattern I've seen and consistently used:

In the main class - for example - `MetricsApplication.java`:

    class Main {
        final MetricRegistry metrics;
        final EventsController eventsController; // This is the object that is instrumented with metrics
    
        Main() {
            metrics = new MetricsRegistry();
            eventsController =
                    new EventsController(metrics);
        }
 
        void run() {
             ....
        }
    }

   
In a class that generates metrics - for example `EventsController.java`
 
    public class EventsController {
        public final Counter networkExceptions;
        public final Counter cloudfrontHits;
        public final Gauge queueSize;
        public final Counter rejectedExecutionExceptions;
        public final Histogram eventSizes;

        public EventsController(MetricRegistry metrics) {
            networkExceptions = metrics.
                counter(name(EventsController.class, "network-exceptions"));
            cloudfrontHits = metrics.
                counter(name(EventsController.class, "cloudfront-hits"));

            queueSize = metrics.
                register(MetricRegistry.name(EventsController.class, "queue-size"),
                (Gauge<Integer>) () -> arrayBlockingQueue.size());

             rejectedExecutionExceptions = metrics.
                counter(name(EventsController.class, "rejected-exceptions"));

            eventSizes = metrics.histogram(name(EventsController.class, "event-sizes"));
        }

This will remove the need for `MetricFactory`, `Metrics` and `Codahale Metrics`. It will also remove the need for registration as that happens right at the beginning as well as locks. AFAIK codahale metrics are thread safe.

The reporter should also be started in the main class. The reporter should filter metrics using a `MetricsFilter`. Example below:

    final SignalFxReporter signalfxReporter = new SignalFxReporter.Builder(
                            metrics,
                            authToken
                    ).setFilter(new MetricFilter())
                     .setMetricMetadata(metricMetadata)
                     .build();

     signalfxReporter.start(60, TimeUnit.SECONDS);

Reply:
1. The main point of creating Metrics interface and Codahale Metrics class is to have loose coupling. You can have other metrics implementation and that can be set as you metrics by setting this "hadoop.cache.data.metrics.class" to your implementation. 

In the implementation you have suggested have direct reference of CodahaleMetrics in the BookKeeper module which we want to avoid. 

2. The reporter is initialized in the initialization of CodahaleMetrics which gets invoked from the main class. The main reason of doing this is the reporters are specific to metrics implementation. For example, we have in this case  JmxReporter from com.codahale.metrics.JmxReporter.